pkg/util/validation/limits.go

```diff
@@ -885,3 +917,93 @@ func EnabledByAnyTenant(tenantIDs []string, f func(string) bool) bool {
 	}
 	return false
 }
+
+// RegisterExtensions registers the extensions type with given name
+// and returns a function to get the extensions from a *Limits instance.
+// The name will be used as YAML/JSON key to decode the extensions.
+func RegisterExtensions[E any](name string) func(*Limits) *E {
+	extensionsType := reflect.TypeOf(new(E))
+	registeredExtensions = append(registeredExtensions, extension{name: name, typ: extensionsType})
```

Comment:
We need to disallow registration of duplicate/conflicting names.

Reply:
Good point. Done in https://github.com/grafana/mimir/pull/4397/commits/772573845223241705f773134cb48be08c2f6fd3

Done it via a panic because I expect extensions to be registered at package initialization level, which would mean that they would still have to panic.
To make it clear, renamed this function to `MustRegisterExtension` in https://github.com/grafana/mimir/pull/4397/commits/c7d7b5b78e8656b66b440a5582f899e661200ea7, if we ever need to return an error, we can define `RegisterExtension`

Reply:
Sounds good. Thanks!